pkg/services/anonymous/anonimpl/impl.go

```diff
@@ -2,178 +2,118 @@ package anonimpl
 	if err != nil {
-		return nil, nil
+		return nil, err
 	}
 
 	return map[string]any{
-		"stats.anonymous.session.count":   anonDeviceCount, // keep session for legacy data
-		"stats.users.device.count":        authedDeviceCount,
 		"stats.anonymous.device.ui.count": anonUIDeviceCount,
-		"stats.users.device.ui.count":     authedUIDeviceCount,
 	}, nil
 }
 
-func (a *AnonDeviceService) untagDevice(ctx context.Context, device *Device) error {
-	key, err := device.Key()
-	if err != nil {
-		return err
-	}
+func (a *AnonDeviceService) tagDeviceUI(ctx context.Context, httpReq *http.Request, device *anonstore.Device) error {
+	key := device.CacheKey()
 
-	if err := a.remoteCache.Delete(ctx, key); err != nil {
-		return err
+	if _, ok := a.localCache.Get(key); ok {
+		return nil
 	}
 
-	return nil
-}
+	a.localCache.SetDefault(key, struct{}{})
 
-func (a *AnonDeviceService) untagUIDevice(ctx context.Context, deviceID string, device *Device) error {
-	key, err := device.UIKey(deviceID)
-	if err != nil {
-		return err
+	if setting.Env == setting.Dev {
+		a.log.Debug("Tagging device for UI", "deviceID", device.DeviceID, "device", device, "key", key)
 	}
 
-	if err := a.remoteCache.Delete(ctx, key); err != nil {
+	if err := a.anonStore.CreateOrUpdateDevice(ctx, device); err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func (a *AnonDeviceService) tagDeviceUI(ctx context.Context, httpReq *http.Request, device Device) error {
-	deviceID := httpReq.Header.Get(deviceIDHeader)
-	if deviceID == "" {
-		return nil
-	}
-
-	if device.Kind == anonymous.AnonDevice {
-		device.Kind = anonymous.AnonDeviceUI
-	} else if device.Kind == anonymous.AuthedDevice {
-		device.Kind = anonymous.AuthedDeviceUI
-	}
-
-	key, err := device.UIKey(deviceID)
-	if err != nil {
-		return err
-	}
-
-	if setting.Env == setting.Dev {
-		a.log.Debug("Tagging device for UI", "deviceID", deviceID, "device", device, "key", key)
-	}
-
-	if _, ok := a.localCache.Get(key); ok {
-		return nil
-	}
-
-	a.localCache.SetDefault(key, struct{}{})
-
-	deviceJSON, err := json.Marshal(device)
+func (a *AnonDeviceService) untagDevice(ctx context.Context,
+	identity *authn.Identity, r *authn.Request, err error) {
 	if err != nil {
-		return err
+		return
 	}
 
-	if err := a.remoteCache.Set(ctx, key, deviceJSON, thirtyDays); err != nil {
-		return err
+	deviceID := r.HTTPRequest.Header.Get(deviceIDHeader)
+	if deviceID == "" {
+		return
 	}
 
-	// remove existing tag when device switches to another kind
-	untagKind := anonymous.AnonDeviceUI
-	if device.Kind == anonymous.AnonDeviceUI {
-		untagKind = anonymous.AuthedDeviceUI
+	errD := a.anonStore.DeleteDevice(ctx, deviceID)
+	if errD != nil {
+		a.log.Debug("Failed to untag device", "error", err)
 	}
-
-	if err := a.untagUIDevice(ctx, deviceID, &Device{
-		Kind:      untagKind,
-		IP:        device.IP,
-		UserAgent: device.UserAgent,
-	}); err != nil {
-		return err
-	}
-
-	return nil
 }
 
+// FIXME: Unexport and remove interface
```

Comment:
If you wan't to un export this you could just pass this as a function to the client :)

Reply:
I kept it as a fixme as I think this PR is already a bit over dimensioned as is :), makes a good beginner 'get to know wire' task